### PR TITLE
disable avx set when gcc is used.

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
GCC on Sandybridge does not support avx2. Even if I use the avx set only in the code, it causes unit test failures. Since we are not able to detect an architecture type, we disable for gcc.

### CODE REVIEW

**Comment by ndellingwood** [comment]
Related to issue #231

**Comment by kyungjoo-kim** [comment]
@ibaned @crtrott Could you merge this to develop branch ?

**Comment by crtrott** [comment]
Guys can you come on the slack channel kokkosteam.slack.com to discuss this?

**Comment by kyungjoo-kim** [comment]
this error only happens on gnu compiler. when intel compiler is fine. so, i don't think that there is possible fix for this except for disabling gnu compiler.

**Comment by ndellingwood** [comment]
Intel 17.0.1 passes:
@kyungjoo-kim @crtrott 

```
[ndellin@apollo TestAllSandia]$ ../../scripts/test_all_sandia intel/17 --build-list=OpenMP --skip-hwloc
Running on machine: apollo
Going to test compilers:  intel/17.0.1
Testing compiler intel/17.0.1
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
#######################################################
PASSED TESTS
#######################################################
intel-17.0.1-OpenMP-release build_time=365 run_time=265
#######################################################
```

**Comment by crtrott** [comment]
Biggest problem with this fix: Intel does set __GNUC__ too!

```c++
-bash-4.1$ cat main.cpp 
#include<cstdio>
int main() {
  
#ifdef __GNUC__ 
  printf("GNU\n");
#else
  printf("NotGNU\n");
#endif
#ifdef __ICC
  printf("Intel\n");
#endif

}
-bash-4.1$ icpc main.cpp 
-bash-4.1$ ./a.out 
GNU
Intel
}
```

So please at least use the KOKKOS compiler macros (i.e. KOKKOS_COMPILER_GNU and KOKKOS_COMPILER_INTEL etc.).

**Comment by crtrott** [comment]
Here is the way to do this compiler specific code:
```c++
-bash-4.1$ cat main.cpp 
#include<Kokkos_Core.hpp>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv);
  {
    #ifdef __GNUC__
       printf("__GNUC__\n");
    #endif
    #ifdef __ICC
       printf("__ICC\n");
    #endif
    #ifdef KOKKOS_COMPILER_GNU
       printf("KOKKOS_COMPILER_GNU\n");
    #endif
    #ifdef KOKKOS_COMPILER_INTEL
       printf("KOKKOS_COMPILER_INTEL\n");
    #endif
  }
  Kokkos::finalize();
}
-bash-4.1$ make CXX=icpc -j KOKKOS_DEVICES=OpenMP KOKKOS_ARCH=SNB
echo "Start Build"
Start Build
-bash-4.1$ ./test.host 
__GNUC__
__ICC
KOKKOS_COMPILER_INTEL
-bash-4.1$ make CXX=icpc -j KOKKOS_DEVICES=OpenMP KOKKOS_ARCH=SNB clean
rm -f Kokkos_Core.o Kokkos_CPUDiscovery.o Kokkos_Error.o Kokkos_ExecPolicy.o Kokkos_HostBarrier.o Kokkos_HostSpace.o Kokkos_HostThreadTeam.o Kokkos_hwloc.o Kokkos_MemoryPool.o Kokkos_Profiling_Interface.o Kokkos_SharedAlloc.o Kokkos_Spinwait.o Kokkos_UnorderedMap_impl.o Kokkos_OpenMP_Exec.o Kokkos_OpenMP_Task.o KokkosCore_config.h KokkosCore_config.tmp libkokkos.a
rm -f *.o *.cuda *.host
-bash-4.1$ make CXX=g++ -j KOKKOS_DEVICES=OpenMP KOKKOS_ARCH=SNB &> out
-bash-4.1$ ./test.host 
__GNUC__
KOKKOS_COMPILER_GNU
```

**Comment by kyungjoo-kim** [comment]
@ndellingwood could you check if this PR branch passes the failed unit tests ? In my workstation (that we reproduced the same failures yeterday), the failures are gone.

**Comment by mhoemmen** [comment]
@micahahoward This may be relevant to the Ifpack2 failed tests you noticed.

**Comment by ndellingwood** [comment]
@kyungjoo-kim running a spot-check on apollo

**Comment by ndellingwood** [comment]
@kyungjoo-kim The same test is still broken with gcc/4.8.4 :(

```
[ RUN      ] openmp.batched_vector_arithmatic_simd_dcomplex2
/home/ndellin/kokkos-kernels/unit_test/../test_common/KokkosKernels_TestUtils.hpp:87: Failure
The difference between double(AT1::abs(val1)) and double(AT2::abs(val2)) is 1.3067744696188346, which exceeds double(AT3::abs(tol)), where
double(AT1::abs(val1)) evaluates to 0,
double(AT2::abs(val2)) evaluates to 1.3067744696188346, and
double(AT3::abs(tol)) evaluates to 0.
//etc.
[  FAILED  ] openmp.batched_vector_arithmatic_simd_dcomplex2 (139 ms)
```

File: src/batched/KokkosBatched_Vector_SIMD.hpp [hunk 1]
```diff
@@ -317,7 +317,7 @@ namespace KokkosBatched {
       }
 
       KOKKOS_INLINE_FUNCTION
-      value_type& operator[](int &i) const {
+      value_type& operator[](const int &i) const {
```
Comment (mhoemmen) [inline src/batched/KokkosBatched_Vector_SIMD.hpp]: Thanks :D

**Comment by kyungjoo-kim** [comment]
@crtrott 

the fix passes spotcheck on apollo (gcc and intel).  if you approve, we can merge it to develop.

**Pull request merged**